RecoParticleFlow/PFProducer/plugins/kdtrees/KDTreeLinkerTrackHcal.cc

```diff
@@ -239,26 +245,85 @@ void KDTreeLinkerTrackHcal::searchLinks() {
 void KDTreeLinkerTrackHcal::updatePFBlockEltWithLinks() {
   //TODO YG : Check if cluster positionREP() is valid ?
 
-  // Here we save in each HCAL cluster the list of phi/eta values of linked clusters (actually tracks).
-  for (BlockElt2BlockEltMap::iterator it = cluster2TargetLinks_.begin(); it != cluster2TargetLinks_.end(); ++it) {
+  // Here we save in each track the list of phi/eta values of linked clusters.
+  for (BlockElt2BlockEltMap::iterator it = target2ClusterLinks_.begin(); it != target2ClusterLinks_.end(); ++it) {
     reco::PFMultiLinksTC multitracks(true);
 
-    for (BlockEltSet::iterator jt = it->second.begin(); jt != it->second.end(); ++jt) {
-      reco::PFRecTrackRef trackref = (*jt)->trackRefPF();
-      const reco::PFTrajectoryPoint& atHCAL = trackref->extrapolatedPoint(trajectoryLayerEntrance_);
-      double tracketa = atHCAL.positionREP().eta();
-      double trackphi = atHCAL.positionREP().phi();
+    //
+    // No restriction on the number of HCAL links per track or T_TO/FROM_DISP or T_FROM_V0/GAMMACON, then fill all links.
+    if (nMaxHcalLinksPerTrack_ < 0. || it->first->trackType(reco::PFBlockElement::T_TO_DISP) ||
+        it->first->trackType(reco::PFBlockElement::T_FROM_DISP) ||
+        it->first->trackType(reco::PFBlockElement::T_FROM_V0) ||
+        it->first->trackType(reco::PFBlockElement::T_FROM_GAMMACONV)) {
+      for (BlockEltSet::iterator jt = it->second.begin(); jt != it->second.end(); ++jt) {
```

Comment:
although you are keeping to the current style of PFBlock, consider a more modern range-based loop, you can save on dereferencing and improve readability:
```C++
for (const auto& cluster : it->second) {
...
}
```

Future PRs to these codes should consider introdicuing range based loops throughout (also done in #31191)

Reply:
instead of the or on trackType (the name should actually be isTrackType) would be more efficient to create a constexpr mask and use that as argument of a new method (say bool isAnyOfTrackType(uint32_t mask);)

Reply:
Thx both. I think I can use isLinkedToDisplacedVertex(). Checking it.